presto-common/src/main/java/com/facebook/presto/common/block/ColumnarRow.java

```diff
@@ -0,0 +1,219 @@
+            averageRowSize += field.getPositionCount() == 0 ? 0 : field.getSizeInBytes() / field.getPositionCount();
+        }
+        return new ColumnarRow(
+                dictionaryBlock,
+                fields,
+                INSTANCE_SIZE + dictionaryBlock.getRetainedSizeInBytes() + sizeOf(dictionaryIds),
+                // The estimated serialized size is the sum of the following:
+                // 1) the offsets size: Integer.BYTES * positionCount. Note that even though ColumnarRow doesn't have the offsets array, the serialized RowBlock still has it. Please see RowBlockEncodingBuffer.
+                // 2) nulls array size: Byte.BYTES * positionCount
+                // 3) the estimated serialized size for the fields Blocks which were just constructed as new DictionaryBlocks:
+                //     the average row size: averageRowSize * the number of rows: nonNullPositionCount
+                (Integer.BYTES + Byte.BYTES) * positionCount + averageRowSize * nonNullPositionCount);
+    }
+
+    private static ColumnarRow toColumnarRow(RunLengthEncodedBlock rleBlock)
+    {
+        Block rleValue = rleBlock.getValue();
+        int positionCount = rleBlock.getPositionCount();
+        ColumnarRow columnarRow = toColumnarRow(rleValue);
+
+        Block[] fields = new Block[columnarRow.getFieldCount()];
+        long averageRowSize = 0;
+        for (int i = 0; i < columnarRow.getFieldCount(); i++) {
+            Block nullSuppressedField = columnarRow.getField(i);
+            if (rleValue.isNull(0)) {
+                // the rle value is a null row so, all null-suppressed fields should empty
+                if (nullSuppressedField.getPositionCount() != 0) {
+                    throw new IllegalArgumentException("Invalid row block");
+                }
+                fields[i] = nullSuppressedField;
+            }
+            else {
+                fields[i] = new RunLengthEncodedBlock(nullSuppressedField, positionCount);
+                averageRowSize += nullSuppressedField.getSizeInBytes() / nullSuppressedField.getPositionCount();
+            }
+        }
+        return new ColumnarRow(
+                rleBlock,
+                fields,
+                INSTANCE_SIZE + rleBlock.getRetainedSizeInBytes(),
+                // The estimated serialized size is the sum of the following:
+                // 1) the offsets size: Integer.BYTES * positionCount. Note that even though ColumnarRow doesn't have the offsets array, the serialized RowBlock still has it. Please see RowBlockEncodingBuffer.
+                // 2) nulls array size: Byte.BYTES * positionCount
+                // 3) the estimated serialized size for the fields Blocks which were just constructed as new RunLengthEncodedBlocks:
+                //     the average row size: averageRowSize * the number of rows: positionCount
+                (Integer.BYTES + Byte.BYTES) * positionCount + averageRowSize * positionCount);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).